linkerd/app/outbound/src/http/logical/policy/route/metrics/labels.rs

```diff
@@ -50,15 +56,55 @@ pub enum Error {
     Unknown,
 }
 
-// === impl Route ===
+// === impl HttpRoute ===
+
+impl HttpRoute {
+    pub fn new(parent: ParentRef, route: RouteRef, host: Option<&str>) -> Self {
+        let host = match host.and_then(|h| url::Host::parse(h).ok()) {
+            Some(url::Host::Domain(s)) => Some(s),
+            // Refrain from emitting label values for url's whose host component are a IPv4
+            // or IPv6 address, to help prevent label cardinality from becoming too noisy.
+            None | Some(url::Host::Ipv4(_) | url::Host::Ipv6(_)) => None,
+        };
```

Comment:
:point_up: this is where we filter out addresses from our labels

Reply:
We already have some prior art in the proxy that avoids the url dependency and is oriented the hyper URI types. Can [`Addr`](https://github.com/linkerd/linkerd2-proxy/blob/89ba5b5d8b39f54de8a8c6e1e8cf552e9c9a5527/linkerd/addr/src/lib.rs#L50) work here?

Reply:
i took a look at the `Addr` types, but these seem subtly different from what we want here. those types are focused on working with the authority, which includes the port in addition to the hostname.

for what it is worth, we do have `url` as a transitive dependency in the workspace today, via the hickory-resolve crate, see: https://github.com/linkerd/linkerd2-proxy/pull/3258/commits/9e3ee3e6611ea96e18d74cf687a33a4c86eb04a8

my chief concern is that using the authority might exacerbate the cardinality problem, as we would have a time series counting e.g. each status code, for each port, for each host. that extra dimension could be costly, until we have some way to limit the number of time series we are emitting.

on the upside, this might help disambiguate between services running on shared infrastructure.

i'm not tremendously opinionated about hostname vs. authority though, and perhaps reusing the `Addr` crate is a nice benefit.

what do you think, @olix0r?

Reply:
Ah, I see what you mean.

I don't think that cardinality is particularly a concern, since parent refs may generally include port information; and disambiguating services by port is pretty natural.

I think the bigger concern is how this compares with values in the TLS case, where we'll have ServerNames that omit ports.

But we still could avoid the url types with, e.g. [`NameAddr::name` ](https://github.com/linkerd/linkerd2-proxy/blob/89ba5b5d8b39f54de8a8c6e1e8cf552e9c9a5527/linkerd/addr/src/lib.rs#L222).

I.e. from an Addr you can get at it via something like `addr.name_addr().and_then(NameAddr::name)`.

This also ensures some amount of validation (i.e. over &str) where we can accept invalid hostnames.

Reply:
i've pushed aed9d303. this commit does two things: (1) avoids a `url` dependency, via use of `linkerd_app_core::dns::Name`, and (2) tweaks the interfaces of `http::logical::policy::route::metrics::labels::HttpRoute` to accept `http` types rather than an `Option<&str>`.

e2447dc0 then reverts our addition of a direct `url` dependency.

this bore out really nicely, thank you for taking the time to work through this! how does this look to you now, @olix0r?

Reply:
marking this as resolved, based on in-air conversation. :slightly_smiling_face: